funnel/views/mixins.py

```diff
@@ -130,3 +133,80 @@ def loader(self, profile, project, section):
             ).first_or_404()
         g.profile = section.project.profile
         return section
+
+
+class DraftViewMixin(object):
+    def get_draft(self, obj=None):
+        """
+        Returns the draft object for `obj`. Defaults to `self.obj`.
+        `obj` is needed in case of multi-model views.
+        """
+        obj = obj if obj is not None else self.obj
+        return Draft.query.get((self.model.__tablename__, obj.uuid))
+
+    def delete_draft(self, obj=None):
+        """
+        Deletes draft for `obj`, or `self.obj` if `obj` is `None`.
+        """
+        draft = self.get_draft(obj)
+        if draft is not None:
+            db.session.delete(draft)
+        else:
+            raise ValueError(_("There is no draft for the given object."))
+
+    def get_draft_data(self, obj=None):
+        """
+        Returns a tuple of the current draft revision and the formdata needed to initialize forms
+        """
+        draft = self.get_draft(obj)
+        if draft is not None:
+            return draft.revision, draft.formdata
+        else:
+            return None, None
+
+    def autosave_post(self, obj=None):
+        """
+        Handles autosave POST requests
+        """
+        obj = obj if obj is not None else self.obj
+        if 'form.revision' not in request.form:
+            # as form.autosave is true, the form should have `form.revision` field even if it's empty
+            return {'error': _("Form must contain a revision ID.")}, 400
+
+        # CSRF check
+        if forms.Form().validate_on_submit():
+            incoming_data = MultiDict(request.form.items(multi=True))
+            client_revision = incoming_data.pop('form.revision')
+            incoming_data.pop('csrf_token', None)
+
+            # find the last draft
+            draft = self.get_draft(obj)
+
+            if draft is not None:
+                if client_revision is None or (client_revision is not None and str(draft.revision) != client_revision):
+                    # draft exists, but the form did not send a revision ID,
+                    # OR revision ID sent by client does not match the last revision ID
+                    return {'error': _("There has been changes to this draft since you last edited it. Please reload.")}, 400
```

Comment:
There _have_ been changes to this item since you last edited it. Please reload this page.